tests/Tokenizer/TokensTest.php

```diff
@@ -1910,6 +1910,34 @@ public function testSettingSizeInTryCatchBlockDoesNotChangeSize(): void
         self::assertSame($size, $tokens->getSize());
     }
 
+    public function testSettingSizeCachePruning(): void
+    {
+        $tokens = Tokens::fromCode('<?php $a = function () {};');
+
+        self::assertSame(11, $tokens->findBlockEnd(Tokens::BLOCK_TYPE_CURLY_BRACE, 10));
+        self::assertSame(10, $tokens->findBlockStart(Tokens::BLOCK_TYPE_CURLY_BRACE, 11));
+        self::assertTrue($tokens->isTokenKindFound(T_FUNCTION));
+        self::assertTrue($tokens->isTokenKindFound('{'));
+        self::assertTrue($tokens->isTokenKindFound('}'));
+        self::assertTrue($tokens->isTokenKindFound(';'));
+        self::assertFalse($tokens->isTokenKindFound(T_CLASS));
+
+        \Closure::bind(static function () use ($tokens): void {
```

Comment:
This looks like overkill. There is a reason `updateSize` is private - to not call it with arbitrary value.

Reply:
I prefer to get it tested because IMO we should be sure that the `updateSize` function manages the cache pruning even if called purely internally/within the Token class only.

I think so as debugging the caching problem (#8228) took me also a day. With this test, we prevent simillar issues in the future.

Reply:
> IMO we should be sure that the `updateSize` function manages the cache pruning even if called purely internally/within the Token class only.

I disagree that we should do that. Method `updateSize ` is private for a reason.

Following your reasoning we should also test all private methods that take an index, when we are sure it exists, and check if it exists, because technically they can be called with `Closure::bind`, the same as `updateSize`.

Reply:
So you want me to remove this test?

Reply:
My view on this is not so generalized.

`private` method can represent some short deduplicated code or self contained functional code that is only not meant to be public. This is the case here. As we not fully expose the code functionality thru the public API, I prefer this to be tested.

However, if you insist on the test removal, I will remove it. Please let me know @kubawerlos. And @Wirone, what is your view on this?

Reply:
I thought https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/8240 was replacing this one.

Reply:
#8240 fixed `Tokens::overrideRange` method.

This PR fixes `Tokens::updateSize` method and optimize Tokens::clearEmptyTokens` method to be mostly `O(1)` which is quite overall improvement as it is called after every fixer is applied.

Reply:
> `private` method can represent some short deduplicated code or self contained functional code that is only not meant to be public.

Of course, but with `Tokens::updateSize` we are in full control that a new size is either 0 or value greater then current/previous size, so there is no chance that there will be a call decreasing the size by 2.

Reply:
We have private method which does something and it should do all the work as the name suggest. This was not the case before this PR, so I fixed this, and added a test.

Reply:
> We have private method which does something and it should do all the work as the name suggest

Once again, this applies to all private methods in the library that has index as parameter and we don't test it exists because it always exists.

We have a method `updateSize` exactly to prevent using the public method `setSize` in the wrong way (e.g. decreasing the size by 2). It does not make sense to test such call.

 If for some reason we want to allow setting arbitrary size from the outside (like in the test) then we can get rid of `updateSize` and stay with `setSize`.

Reply:
So tell me how should I amend this PR to pass your review.

Reply:
With some test of real-world usage, like it is in https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/8240

Reply:
Of course I have tried to find a way to test this from public API.

However, the fixed `updateSize` code only moved some cache pruning code into this method (l1236-1245) and there is no way how to code a before failing test. The only way to assert the coherent behaviour of this method is to test the private/internal method.

So please tell me if this PR is good as it is or if you have any concrete ideas how to improve it.

Reply:
> The only way to assert the coherent behaviour of this method is to test the private/internal method.

In other words: "there is no way to break it using public API", right?

> So please tell me if this PR is good as it is or if you have any concrete ideas how to improve it.

I can think of 3rd option 😉

Reply:
> In other words: "there is no way to break it using public API", right?

Correct.

> I can think of 3rd option 😉

What is the 3rd option?

Reply:
Closing the PR until there is a valid reason to make the changes (maybe some benchmarks that anyone can run locally to verify).

Reply:
I see no such improvement.

Reply:
see first change in `clearEmptyTokens`

Reply:
> see first change in `clearEmptyTokens` and PR title

Early return for a happy path is no proof for changing O(N) to O(1), neither is PR title.

Reply:
Why? The loop after is normally fully iterated.

Reply:
> Why?

Because that's not how Big O Notation works: [A Beginner's Guide to The Big O Notation](https://hackernoon.com/a-beginners-guide-to-the-big-o-notation-yb7332wf).

Reply:
My view on this is:
- my optimization is O(1) as when there are no empty tokens\* the early return is used which is conditioned on constant time expression
- before/the loop after is O(N) as it scales lineary with the tokens size as long as there are no empty tokens\*

\* no empty tokens - almost always (and definitively always on already fixed code)

What am I missing?

Reply:
> * no empty tokens - almost always (and definitively always on already fixed code)

We don't call `Tokens::clearEmptyTokens` on already fixed code... we call it only when token's collection is changed.

> My view on this

... does not change how Big O Notation works.

I feel this is going nowhere. I'm not going to be answering here anymore. Have a happy rest of the weekend 🍰.

Reply:
> ... does not change how Big O Notation works.
> 
> I feel this is going nowhere.

So would you tell me how it works to be able to meet on the same level?